### PR TITLE
Feature/dataframejoin

### PR DESCRIPTION
**Pull request by martinfalk**
utility functions and processor for joining two DataFrames

### CODE REVIEW

**Review by petersteneteg — DISMISSED** [review]

File: modules/dataframe/include/inviwo/dataframe/datastructures/column.h [hunk 1]
```diff
@@ -94,33 +94,42 @@ class TemplateColumn : public Column {
...
-    virtual void set(size_t idx, const T &value);
+    virtual void add(const std::string& value) override;
+
+    /**
+     * \brief appends all rows from column \p col
+     *
+     * @param col
+     * @throws Exception if data format does not match
+     */
+    virtual void append(const Column& col) override;
```
Comment (ResearchDaniel) [inline modules/dataframe/include/inviwo/dataframe/datastructures/column.h:133]: Documentation should be in base class and copied here using doxygen command?

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#include <inviwo/dataframe/dataframemoduledefine.h>
+
+#include <inviwo/core/common/inviwo.h>
+#include <inviwo/core/datastructures/buffer/buffer.h>
+#include <inviwo/core/io/serialization/serializable.h>
+#include <inviwo/dataframe/datastructures/dataframe.h>
+
+namespace inviwo {
+
+namespace dataframeutil {
```
Comment (ResearchDaniel) [inline modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h]: An option would be to just have dataframe as namespace. Breaking change but moving the file will be a breaking change anyway

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+void IVW_MODULE_DATAFRAME_API copyBufferRange(std::shared_ptr<const BufferBase> src,
+                                              std::shared_ptr<BufferBase> dst, ivec2 range,
+                                              size_t dstStart = 0);
+
+/**
+ * \brief create a new DataFrame by appending the columns of DataFrame \p right to DataFrame \p left
+ *
+ * @param ignoreDuplicates duplicate columns, i.e. same column header, are ignored if true
+ * @param fillMissingRows  if true, missing rows in either DataFrame are filled with 0 or
+ *                         "undefined" (for categorical columns)
```
Comment (ResearchDaniel) [inline modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: Should we use NaN instead on 0? Or force specification of value to fill in?
Forcing 0 can cause issues later

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+std::shared_ptr<DataFrame> IVW_MODULE_DATAFRAME_API appendColumns(const DataFrame& left,
+                                                                  const DataFrame& right,
+                                                                  bool ignoreDuplicates = false,
+                                                                  bool fillMissingRows = false);
+
+/**
+ * \brief create a new DataFrame by appending the rows of DataFrame \p bottom to DataFrame \p top
+ *
+ * @param matchByName    if true, column headers are used for matching columns. Otherwise columns
+ *                       are matched by order (default)
```
Comment (ResearchDaniel) [inline modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:67]: Does it not make more sense to match by name by default?

File: modules/dataframe/src/util/dataframeutil.cpp [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+#include <inviwo/core/util/zip.h>
+#include <inviwo/core/util/assertion.h>
+
+#include <fmt/format.h>
+
+namespace inviwo {
+
+namespace dataframeutil {
+
+std::shared_ptr<BufferBase> cloneBufferRange(std::shared_ptr<const BufferBase> buffer,
```
Comment (ResearchDaniel) [inline modules/dataframe/src/util/dataframeutil.cpp:44]: Not part of this PR but we have placed this type in buffer/algorithm/ for other type (volume/image)

**Review by petersteneteg — DISMISSED** [review]

File: modules/dataframe/src/util/dataframeutil.cpp [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+    addColumns(left, rows.first, false);
+    addColumns(right, rows.second, true);
+
+    dataframe->updateIndexBuffer();
+    return dataframe;
+}
+
+std::shared_ptr<DataFrame> combineDataFrames(std::vector<std::shared_ptr<DataFrame>> dataFrames,
+                                             bool skipIndexColumn, std::string skipcol) {
```
Comment (ResearchDaniel) [inline modules/dataframe/src/util/dataframeutil.cpp:226]: Simply combine(...) because it is in the namespace?

**Review by ResearchDaniel — DISMISSED** [review]

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 3 above)
Comment (martinfalk) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: Not sure. At the moment we definitely don't have a proper way to encode "missing" or empty data. In Sql it's `NULL`. `"undefined"` for categorical columns is already stretching it a bit.

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 4 above)
Comment (martinfalk) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:67]: wouldn't say so. Common scenario from my perspective is to merge two dataframes from different sources, i.e. different headers.

File: modules/dataframe/src/util/dataframeutil.cpp (hunk 6 above)
Comment (martinfalk) [reply modules/dataframe/src/util/dataframeutil.cpp:226]: that was pre-existing prior this PR (similar to `cloneBufferRange`)

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 3 above)
Comment (ResearchDaniel) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: I think that allowing values to fill in would be the way to go. The default could be NaN for numerical and undefined for categorical though. Then it is kind of up to the developer to choose 0 if that is more appropriate

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 4 above)
Comment (ResearchDaniel) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:67]: Ok, we had another scenario where the columns were moved, but I guess either is valid

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 3 above)
Comment (martinfalk) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: NaN does not exist for integral types though.

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 2 above)
Comment (martinfalk) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h]: done. Also mentioned in changelog.

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 3 above)
Comment (petersteneteg) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: One should definitely not use NaN. It's not supposed to represent missing data, it is the result it a failed computation.
if you want to have a state for missing you have to provide that externally as in using `optional<int>` or `optional<float>`

Comment (petersteneteg) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: But having the data stored in buffers will not work. Hence it you want to have missing data in a data frame it needs to be dramatically redesigned.

**Review by petersteneteg — APPROVED** [review]

File: modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h (hunk 3 above)
Comment (ResearchDaniel) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: One should not use 0 to represent missing data either. Let the developer decide at least

Comment (petersteneteg) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: The thing is the current data frame can NOT represent missing data, there is nothing to decide.

Comment (ResearchDaniel) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: With that reasoning there should be an exception thrown and that is not really an option either.

Comment (petersteneteg) [reply modules/dataframe/include/inviwo/dataframe/util/dataframeutil.h:54]: there should only be an inner join. we can't support any other one.
